src/protocol.rs

```diff
@@ -339,9 +341,47 @@ impl ParticipantIdentifier {
     }
 }
 
-impl ProofContext for Vec<ParticipantIdentifier> {
+/// The SharedContext contains fixed known parameters accross the entire
+/// protocol. It does not however contain the entire protocol context.
+pub struct SharedContext {
+    pub participants: Vec<ParticipantIdentifier>,
+    pub generator: CurvePoint,
+    pub order: BigNumber,
+}
+impl ProofContext for SharedContext {
     fn as_bytes(&self) -> Vec<u8> {
-        self.iter().flat_map(|pid| pid.0.to_le_bytes()).collect()
+        [
+            self.participants
+                .iter()
+                .flat_map(|pid| pid.0.to_le_bytes())
+                .collect(),
+            bincode::serialize(&self.generator).unwrap(),
```

Comment:
todo: this is kind of safe because generator is constant, but I don't really like the unwrap here. Can you update this function to return a result and throw an error (probably `InternalInvariantFailed` because it shouldn't happen) instead?